fvm/src/kernel/default.rs

```diff
@@ -342,18 +363,29 @@ where
         Ok(k)
     }
 
-    fn block_read(&self, id: BlockId, offset: u32, buf: &mut [u8]) -> Result<u32> {
+    fn block_read(&mut self, id: BlockId, offset: u32, buf: &mut [u8]) -> Result<u32> {
+        self.call_manager
+            .charge_gas(self.call_manager.price_list().on_block_read_base())?;
+
         let data = self.blocks.get(id).or_illegal_argument()?.data();
-        Ok(if offset as usize >= data.len() {
+
+        let len = if offset as usize >= data.len() {
             0
         } else {
-            let len = buf.len().min(data.len());
-            buf.copy_from_slice(&data[offset as usize..][..len]);
-            len as u32
-        })
+            let copy_len = buf.len().min(data.len());
+            buf.copy_from_slice(&data[offset as usize..][..copy_len]);
+            copy_len
+        };
+
+        self.call_manager
+            .charge_gas(self.call_manager.price_list().on_block_read_per_byte(len))
+            .map(|_| len as u32)
```

Comment:
Unlike `open`, we _should_ charge this up-front. I mean, there's no observable difference, but we save ourselves a bit of work if the gas doesn't cover the copy.